### DISCUSSION TITLE
Password hashing functions for RabbitMq 3.13.7

### DISCUSSION THREAD

**Discussion by ElenaShlykova** [rabbitmq/rabbitmq-server#12747] [Questions] [answered]
### Community Support Policy

- [X] I have read [RabbitMQ's Community Support Policy](https://github.com/rabbitmq/rabbitmq-server/blob/main/COMMUNITY_SUPPORT.md)
- [X] I agree to provide all relevant information (versions, logs, rabbitmq-diagnostics output, detailed reproduction steps)

### RabbitMQ version used

3.13.7 or older

### Erlang version used

26.2.x

### Operating system (distribution) used

Windows Server 2019

### How is RabbitMQ deployed?

Windows installer

### What problem are you trying to solve?

Our cybersecurity team raised a question about how credentials and passwords are stored in the rabbitmq 3.13.7.
As per doucmentation we found the SHA256 is used by default: https://www.rabbitmq.com/docs/3.13/passwords
However we did not found defailts if a salt is provided? If so, how is randomness ensured?

We would be grateful for the details provided.

**Comment by lukebakken** [comment, marked as answer]
> However we did not found defailts if a salt is provided? If so, how is randomness ensured?

It is clearly documented on our website:

https://www.rabbitmq.com/docs/passwords#this-is-the-algorithm

RabbitMQ is open-source, and you have the ability to see _exactly_ how passwords are dealt with:

https://github.com/search?q=repo%3Arabbitmq%2Frabbitmq-server+Salt&type=code

https://github.com/rabbitmq/rabbitmq-server/blob/047cc5a08409ddd43d73c6f5076130277d410462/deps/rabbit_common/src/rabbit_password.erl#L27-L29

https://www.erlang.org/doc/apps/stdlib/rand.html#uniform/1

If you need to see how `rand:uniform/1` is implemented in Erlang:

https://github.com/search?q=repo%3Aerlang%2Fotp%20uniform&type=code

https://github.com/erlang/otp/blob/master/lib/crypto/c_src/rand.c#L85-L135
